compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/QueryDesugar.java

```diff
@@ -0,0 +1,250 @@
+        this.names = Names.getInstance(context);
+        this.types = Types.getInstance(context);
+        this.dlog = BLangDiagnosticLog.getInstance(context);
+        this.desugar = Desugar.getInstance(context);
+        this.anonymousModelHelper = BLangAnonymousModelHelper.getInstance(context);
+    }
+
+    public static QueryDesugar getInstance(CompilerContext context) {
+        QueryDesugar desugar = context.get(QUERY_DESUGAR_KEY);
+        if (desugar == null) {
+            desugar = new QueryDesugar(context);
+        }
+
+        return desugar;
+    }
+
+    BLangStatementExpression desugarQueryExpr(BLangQueryExpr queryExpr, SymbolEnv env) {
+        List<BLangFromClause> fromClauseList = queryExpr.fromClauseList;
+        BLangFromClause fromClause = fromClauseList.get(0);
+        BLangSelectClause selectClause = queryExpr.selectClause;
+        List<BLangWhereClause> whereClauseList = queryExpr.whereClauseList;
+        DiagnosticPos pos = fromClause.pos;
+
+        // Create Foreach statement
+        //
+        // Below query expression :
+        //      from var person in personList
+        //
+        // changes as,
+        //      foreach var person in personList {
+        //          ....
+        //      }
+        BLangForeach leafForeach = buildFromClauseBlock(fromClauseList);
+        BLangBlockStmt foreachBody = ASTBuilderUtil.createBlockStmt(pos);
+
+        BType outputArrayType;
+        if (selectClause.expression != null && selectClause.expression.type != null) {
+            outputArrayType = new BArrayType(selectClause.expression.type);
+        } else {
+            outputArrayType = fromClause.varType;
+        }
+
+        BLangListConstructorExpr emptyArrayExpr = ASTBuilderUtil.createEmptyArrayLiteral(pos,
+                (BArrayType) outputArrayType);
+        BVarSymbol emptyArrayVarSymbol = new BVarSymbol(0, new Name("$outputDataArray$"),
+                env.scope.owner.pkgID, outputArrayType, env.scope.owner);
+        BLangSimpleVariable outputArrayVariable =
+                ASTBuilderUtil.createVariable(pos, "$outputDataArray$", outputArrayType,
+                        emptyArrayExpr, emptyArrayVarSymbol);
+
+        // Create temp array variable
+        //      Person[] x = [];
+
+        BLangSimpleVariableDef outputVariableDef =
+                ASTBuilderUtil.createVariableDef(pos, outputArrayVariable);
+        BLangSimpleVarRef outputVarRef = ASTBuilderUtil.createVariableRef(pos, outputArrayVariable.symbol);
+
+        // Create indexed based access expression statement
+        //      x[x.length()] = {
+        //         firstName: person.firstName,
+        //         lastName: person.lastName
+        //      };
+
+        if (selectClause.expression.type == null) {
+            selectClause.expression.type = fromClause.varType;
+        }
+
+        BLangInvocation lengthInvocation = createLengthInvocation(selectClause.pos, outputArrayVariable.symbol);
```

Comment:
Shall we replace this with a simple array:push function?

Reply:
Facing an issue for primitive arrays @ runtime when using push method. Will attend this separately.